packages/next/pages/_app.tsx

```diff
@@ -31,15 +32,15 @@ async function appGetInitialProps({
   return { pageProps }
 }
 
-export default class App<P = any, CP = {}, S = {}> extends React.Component<
-  P & AppProps<CP>,
-  S
-> {
+export default class App<
+  P extends AppInitialProps = AppInitialProps,
+  S = {}
+> extends React.Component<AppProps<P>, S> {
   static origGetInitialProps = appGetInitialProps
   static getInitialProps = appGetInitialProps
 
   render() {
-    const { Component, pageProps } = this.props as AppProps<CP>
+    const { Component, pageProps } = this.props
```

Comment:
This way also allows us to remove this "hack" type cast, types now behave naturally here